### DISCUSSION TITLE
Torch dependency & Callback functionality

### DISCUSSION THREAD

**Discussion by tortka** [Evolutionary-Intelligence/pypop#181] [General]
Hello PyPop7-Devs,

I'm the main developer of pyanno4rt, a package for radiotherapy dose optimization, and I've tested a few algorithms from your package. Really interesting selection of algorithms, and even though my optimization problems are quite high-dimensional (>5000 variables), the computational speed seems pretty good, which is why I'm seriously thinking about integrating your package as a dependency in the first release of pyanno4rt! However, currently there's two major pitfalls for me, maybe you got a solution here:

1. The installation of PyPop7 comes with large memory requirements, especially due to the torch dependency and the NVIDIA-GPU supports. Actually, I'm not even utilizing GPUs, and therefore it would be best to save the memory space (also keeping in mind that a package with 3.5 GB of dependencies is more attractive than one with 7 GB). Is there any option to have torch as an optional dependency and thus disable its installation? (Sure, I could remove it afterwards, but that's neither practical nor reasonable, I believe).
2. I'm missing a callback functionality to retrieve information from the optimization iterations, e.g. for logging and displaying in a graphical user interface. Did I just miss that feature, or could you add it otherwise? Right now, I get message prints on the standard output, which is fine, but in my application, I would need to redirect these messages to other streams.

Thanks for an answer and for your impressive work!

Regards,
Tim

**Comment by Evolutionary-Intelligence** [comment]
@TimOrtkamp Thanks very much for your kindly suggestions. We have removed **torch** for the latest version v0.0.79. Now the memory requirements have been significantly reduced, in order to better meet your application requirements.

In the future, we will add the *seemingly necessary* **callback** functionality, according to your good suggestion.

**Comment by tortka** [comment]
@Evolutionary-Intelligence Fantastic! I've spend the day re-integrating your package into mine, and now it is running as desired :) I've added references to your documentation page and preprint within the respective wrapper files. Also, I'm looking forward to the addition of a callback functionality! Maybe a follow-up question: from the algorithms implemented in PyPop7, are there specific ones you would recommend in case of a high-dimensional (>5000 variables), nonlinear and partially non-convex optimization problem. So far, I've made some good experiences with LMCMA and LMMAES, but it would be interesting to know if there's any further options to test (without having to go through all available large-scale optimization algorithms) ;)
Best, Tim
